mm2src/lp_ordermatch/orderbook_rpc.rs

```diff
@@ -178,3 +210,154 @@ pub async fn orderbook_rpc(ctx: MmArc, req: Json) -> Result<Response<Vec<u8>>, S
     let response = try_s!(json::to_vec(&response));
     Ok(try_s!(Response::builder().body(response)))
 }
+
+#[derive(Debug, Display, Serialize, SerializeErrorType)]
+#[serde(tag = "error_type", content = "error_data")]
+pub enum OrderbookRpcError {
+    BaseRelSame,
+    BaseRelSameOrderbookTickersAndProtocols,
+    CoinConfigNotFound(String),
+    CoinIsWalletOnly(String),
+    P2PSubscribeError(String),
+}
+
+impl HttpStatusCode for OrderbookRpcError {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            OrderbookRpcError::BaseRelSame
+            | OrderbookRpcError::BaseRelSameOrderbookTickersAndProtocols
+            | OrderbookRpcError::CoinConfigNotFound(_)
+            | OrderbookRpcError::CoinIsWalletOnly(_) => StatusCode::BAD_REQUEST,
+            OrderbookRpcError::P2PSubscribeError(_) => StatusCode::INTERNAL_SERVER_ERROR,
+        }
+    }
+}
+
+#[derive(Serialize)]
+pub struct OrderbookV2Response {
+    asks: Vec<AggregatedOrderbookEntryV2>,
+    base: String,
+    bids: Vec<AggregatedOrderbookEntryV2>,
+    net_id: u16,
+    num_asks: usize,
+    num_bids: usize,
+    rel: String,
+    timestamp: u64,
+    total_asks_base_vol: MmNumberMultiRepr,
+    total_asks_rel_vol: MmNumberMultiRepr,
+    total_bids_base_vol: MmNumberMultiRepr,
+    total_bids_rel_vol: MmNumberMultiRepr,
+}
+
+pub async fn orderbook_rpc_v2(
+    ctx: MmArc,
+    req: OrderbookReq,
+) -> Result<OrderbookV2Response, MmError<OrderbookRpcError>> {
+    if req.base == req.rel {
+        return MmError::err(OrderbookRpcError::BaseRelSame);
+    }
+    let base_coin_conf = coin_conf(&ctx, &req.base);
+    if base_coin_conf.is_null() {
+        return MmError::err(OrderbookRpcError::CoinConfigNotFound(req.base));
+    }
+    if is_wallet_only_conf(&base_coin_conf) {
+        return MmError::err(OrderbookRpcError::CoinIsWalletOnly(req.base));
+    }
+    let rel_coin_conf = coin_conf(&ctx, &req.rel);
+    if rel_coin_conf.is_null() {
+        return MmError::err(OrderbookRpcError::CoinConfigNotFound(req.rel));
+    }
+    if is_wallet_only_conf(&rel_coin_conf) {
+        return MmError::err(OrderbookRpcError::CoinIsWalletOnly(req.rel));
+    }
```

Comment:
I think this code could be simplified - the function is already quite big and hard to read, is it possible to refactor it into different functions?

Reply:
Good point, `fn get_tradeable_coin_conf(...)` should be a good match for this.

Reply:
@Milerius Done.